### PR TITLE
use psi based type reference to defer type resolution

### PR DESCRIPTION
**Pull request by neetopia**

### CODE REVIEW

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationJavaImpl.kt [hunk 1]
```diff
@@ -27,7 +28,7 @@ class KSPropertyDeclarationJavaImpl private constructor(private val ktJavaFieldS
         get() = null
 
     override val type: KSTypeReference by lazy {
-        KSTypeReferenceImpl.getCached(ktJavaFieldSymbol.returnType, this@KSPropertyDeclarationJavaImpl)
+        KSTypeReferenceResolvedImpl.getCached(ktJavaFieldSymbol.returnType, this@KSPropertyDeclarationJavaImpl)
```
Comment (ting-yuan) [inline kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationJavaImpl.kt:31]: Ditto. Needs to be `KSTypeReferenceImpl` when this is backed by sources.

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSCallableReferenceImpl.kt [hunk 2]
```diff
@@ -15,18 +17,18 @@ class KSCallableReferenceImpl private constructor(
             cache.getOrPut(IdKeyPair(ktFunctionalType, parent)) { KSCallableReferenceImpl(ktFunctionalType, parent) }
     }
     override val receiverType: KSTypeReference?
-        get() = ktFunctionalType.receiverType?.let { KSTypeReferenceImpl.getCached(it) }
+        get() = ktFunctionalType.receiverType?.let { KSTypeReferenceResolvedImpl.getCached(it) }
```
[the remark this answers is not in the document]
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSCallableReferenceImpl.kt:20]: KtCallableReference is left as TODO in this PR.

File: kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationJavaImpl.kt (hunk 1 above)
Comment (neetopia) [reply kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSPropertyDeclarationJavaImpl.kt:31]: Java uses `PsiType`s which are yet to be supported in analysis API to convert to `KtType`s

**Comment by ting-yuan** [comment]
`KSTypeReferenceImpl.defer()` needs to be implemented. Other than that LGTM.

**Review by ting-yuan — APPROVED** [review]

**Pull request merged**